### PR TITLE
Feature/switch to postgresql

### PR DESCRIPTION
**Pull request by crudenesss**
Implemented switch from **MongoDB** to **PostgreSQL**. Used SQLAlchemy ORM.

### CODE REVIEW

File: env_example [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Non privileged user credentials (to execute commands with)
+POSTGRES_USERNAME=user
+POSTGRES_PASSWORD=password
+
+# In-app admin credentials
+APP_ADMINUSER=admin
```
Comment (KernelKrise) [inline env_example:6]: Please, separate words using "_"

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+      - app-network
+
+  app:
+    container_name: yapp-space
+    build:
+      context: .
+    image: crudenesss/yapp-space
+    depends_on:
+      postgres:
+        condition: service_started
```
Comment (KernelKrise) [inline docker-compose.yml:23]: Please, add healthcheck

File: setup-scripts/setup.sh [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+psql -c "\c ${PGDATABASE};"
+
+# Create all tables 
+psql -c "CREATE TABLE roles ( 
+    role_id VARCHAR(32) PRIMARY KEY,
+    role_name VARCHAR(64)
+);"
+
+psql -c "CREATE TABLE users (
+    user_id VARCHAR(80) PRIMARY KEY,
```
Comment (KernelKrise) [inline setup-scripts/setup.sh:14]: uuid4 length is 36, please decrease varchar length to appropriate

File: setup-scripts/setup.sh [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    username VARCHAR(32) NOT NULL,
+    passwd VARCHAR(256) NOT NULL,
+    email VARCHAR(320) NOT NULL,
+    bio VARCHAR(256),
+    profile_picture VARCHAR(128),
```
Comment (KernelKrise) [inline setup-scripts/setup.sh:19]: Please, decrease length to appropriate

File: setup-scripts/setup.sh [hunk 5]
```diff
@@ -0,0 +1,52 @@
+#!/bin/bash
```
Comment (KernelKrise) [inline setup-scripts/setup.sh:1]: Please, minimise column type sizes

File: flask_chat/services.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        """
+        try:
+            session = self.session()
+
+            if kwargs:
+                results = session.query(User).filter_by(**kwargs).all()
+            else:
+                results = session.query(User).all()
+
+            logger.debug(results)
```
Comment (KernelKrise) [inline flask_chat/services.py:71]: Please, delete redundant log

File: flask_chat/services.py [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+            session.close()
+            return results
+        except SQLAlchemyError:
```
Comment (KernelKrise) [inline flask_chat/services.py:74]: Please, think about closing session on error. Here and in all such cases

File: flask_chat/services.py [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+
+        ### Returns:
+            _bool_:
+            _True_ if operation is successful, otherwise _False_.
+        """
+        try:
+            session = self.session()
+
+            if session.query(User).all():
+                role = "user"
```
Comment (KernelKrise) [inline flask_chat/services.py:99]: Here. and in all other cases, use variables with role names

File: flask_chat/services.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+            else:
+                role = "admin"
+
+            new_user = User(
+                user_id="uid-" + random_strings_generator(),
```
Comment (KernelKrise) [inline flask_chat/services.py:104]: Please, delete redundant "uid-"

File: flask_chat/services.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+            logger.debug("%s rows retrieved: %s", len(result), result)
+            session.close()
+        except SQLAlchemyError:
+            logger.error("An error occured while establishing conection with PostgreSQL instance.")
+            return None
+
+        if jsonify:
+            for index, message in enumerate(result):
+                result[index] = message._mapping
+                logger.debug(result[index])
```
Comment (KernelKrise) [inline flask_chat/services.py:268]: Please, delete redundant log

File: flask_chat/views.py [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+    [user_data] = user_service.get_user_info(username=username)
+    if not user_data:
+        logger.error("Unable to load user info.")
+        return render_template(
+            "error.html",
+            message="Unable to process info about your identity. Pleasy try later.",
+        )
+
+    # Here goes verifying password hashes
+    if not user_data.verify_password(password):
```
Comment (KernelKrise) [inline flask_chat/views.py:162]: [CRITICAL] Please, combine this if with password verification. Validate what DB return if user not exists. If something wrong, return message should be the same all the time

File: flask_chat/views.py [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+
+    # Check if request contains data about file uploaded
+    if request.files["file"].read() != b"":
+        content = request.files["file"]
+        content.seek(0)
+        content_clean = content.read()
+        content.seek(0)
+
+        # Check if received file's content is empty or if file was not chosen
+        if content_clean == b"":
```
Comment (KernelKrise) [inline flask_chat/views.py:250]: Delete redundant if. Emptyness check done before

File: flask_chat/views.py [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+            "profile.html",
+            form=form,
+            current_user=current_user,
+            csrf_token=csrf_token,
+            data=user_data,
+            web_name=WEBSITE_NAME,
+        )
+
+    # Check if request contains data about file uploaded
+    if request.files["file"].read() != b"":
```
Comment (KernelKrise) [inline flask_chat/views.py:243]: Please, delete redundant code with seek() and double read of file. Just read file into variable, than do if

File: flask_chat/views.py [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+            content.save(file)
+
+        # remove old picture to avoid trashing
+        if user_data.get("profile_picture") is not None:
+            os.remove(
+                os.path.join(PROFILE_PICTURE_STORAGE_PATH, user_data.get("profile_picture"))
+            )
+            logger.debug("previous picture is removed")
+
+        # Update profile picture filename in database
```
Comment (KernelKrise) [inline flask_chat/views.py:274]: Move this code up to avoid situation, when image deleted, but db update failed

File: flask_chat/views.py [hunk 15]
```diff
@@ -0,0 +1,414 @@
...
+
+    if not form.validate():
+        logger.debug("Validation of profile edit form's input failed")
+        return redirect(request.url)
+
+    # Create dictionary for values that were changed
+    newvalues = {}
+
+    for field in form:
+
```
Comment (KernelKrise) [inline flask_chat/views.py:298]: Please, add whitelist of allowed field names

File: flask_chat/views.py [hunk 16]
```diff
@@ -0,0 +1,414 @@
...
+    profile_picture_name = user_info[0].profile_picture
+
+    # Check if user has his profile picture set -
+    # if not - return path with default picture
+    if profile_picture_name is None:
+        logger.debug("No picture to retrieve: set to default")
+        return send_file(DEFAULT_PROFILE_PICTURE_PATH)
+    else:
+        return send_file(
```
Comment (KernelKrise) [inline flask_chat/views.py:402]: Check if file exists

File: env_example (hunk 1 above)
Comment (crudenesss) [reply env_example]: Changed to more readable variable names.

File: docker-compose.yml (hunk 2 above)
Comment (crudenesss) [reply docker-compose.yml]: Added healthcheck.

File: flask_chat/views.py (hunk 16 above)
Comment (crudenesss) [reply flask_chat/views.py]: Added additional check.

File: flask_chat/views.py (hunk 15 above)
Comment (crudenesss) [reply flask_chat/views.py:286]: Added whitelist of allowed fields.

File: flask_chat/views.py (hunk 14 above)
Comment (crudenesss) [reply flask_chat/views.py:253]: Fixed.

File: flask_chat/views.py (hunk 13 above)
Comment (crudenesss) [reply flask_chat/views.py]: Fixed.

File: flask_chat/views.py (hunk 12 above)
Comment (crudenesss) [reply flask_chat/views.py]: Fixed.

File: flask_chat/views.py (hunk 11 above)
Comment (crudenesss) [reply flask_chat/views.py]: Fixed.

File: flask_chat/services.py (hunk 10 above)
Comment (crudenesss) [reply flask_chat/services.py]: Removed redundant log.

File: flask_chat/services.py (hunk 7 above)
Comment (crudenesss) [reply flask_chat/services.py:77]: Fixed.

File: flask_chat/services.py (hunk 6 above)
Comment (crudenesss) [reply flask_chat/services.py]: Removed redundant log.

**Comment by KernelKrise** [comment]
Internal Server Error on user info changes

**Comment by KernelKrise** [comment]
Internal server error on /manage

**Comment by KernelKrise** [comment]
Please, add "Username or Email" in /login, not just Username (html)

**Comment by KernelKrise** [comment]
Can't login with email on /login

**Comment by crudenesss** [comment]
> Please, add "Username or Email" in /login, not just Username (html)
> Can't login with email on /login

I haven't implemented this feature yet, otherwise i would reference it in issue #7

**Comment by KernelKrise** [comment]
Delete "Switch to personal view" button, when viewing profile of different user

**Comment by KernelKrise** [comment]
Please, remove redundant admin info from env_example

**Pull request merged**